tgui/packages/tgui/index.js

```diff
@@ -0,0 +1,218 @@
+import 'core-js/web/queue-microtask';
+import 'core-js/web/timers';
+import 'regenerator-runtime/runtime';
+import './polyfills/html5shiv';
+import './polyfills/ie8';
+import './polyfills/dom4';
+import './polyfills/css-om';
+import './polyfills/inferno';
+
+// Themes
+import './styles/main.scss';
+import './styles/themes/abductor.scss';
+import './styles/themes/cardtable.scss';
+import './styles/themes/hackerman.scss';
+import './styles/themes/malfunction.scss';
+import './styles/themes/ntos.scss';
+import './styles/themes/paper.scss';
+import './styles/themes/retro.scss';
+import './styles/themes/syndicate.scss';
+
+import { perf } from 'common/perf';
+import { render } from 'inferno';
+import { setupHotReloading } from 'tgui-dev-server/link/client';
+import { loadCSS } from './assets';
+import { backendUpdate, backendSuspendSuccess, selectBackend, sendMessage } from './backend';
+import { setupDrag } from './drag';
+import { logger } from './logging';
+import { createStore, StoreProvider } from './store';
+
+perf.mark('inception', window.__inception__);
+perf.mark('init');
+
+const store = createStore();
+let reactRoot;
+let initialRender = true;
+
+const renderLayout = () => {
+  perf.mark('render/start');
+  const state = store.getState();
+  const { suspended, assets } = selectBackend(state);
+  // Initial render setup
+  if (initialRender) {
+    logger.log('initial render', state);
+    // Setup dragging
+    if (initialRender !== 'recycled') {
+      setupDrag();
+    }
+  }
+  // Start rendering
+  const { getRoutedComponent } = require('./routes');
+  const Component = getRoutedComponent(state);
+  const element = (
+    <StoreProvider store={store}>
+      <Component />
+    </StoreProvider>
+  );
+  if (!reactRoot) {
+    reactRoot = document.getElementById('react-root');
+  }
+  render(element, reactRoot);
+  if (suspended) {
+    return;
+  }
+  perf.mark('render/finish');
+  // Report rendering time
+  if (process.env.NODE_ENV !== 'production') {
+    if (initialRender === 'recycled') {
+      logger.log('rendered in',
+        perf.measure('render/start', 'render/finish'));
+    }
+    else if (initialRender) {
+      logger.debug('serving from:', location.href);
+      logger.debug('bundle entered in',
+        perf.measure('inception', 'init'));
+      logger.debug('initialized in',
+        perf.measure('init', 'render/start'));
+      logger.log('rendered in',
+        perf.measure('render/start', 'render/finish'));
+      logger.log('fully loaded in',
+        perf.measure('inception', 'render/finish'));
+    }
+    else {
+      logger.debug('rendered in',
+        perf.measure('render/start', 'render/finish'));
+    }
+  }
+  if (initialRender) {
+    initialRender = false;
+  }
+  // Load assets
+  assets?.styles?.forEach(filename => loadCSS(filename));
+};
+
+// Parse JSON and report all abnormal JSON strings coming from BYOND
+const parseStateJson = json => {
+  let reviver = (key, value) => {
```

Comment:
Minor as all heck, but I think this can be pulled out of this function rather than redefined every call.

Reply:
Good call, might also improve the memory leak situation on IE8.